src/scalar.rs

```diff
@@ -0,0 +1,137 @@
+    /// Scalar representing `1`
+    pub const ONE: Scalar = Scalar([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]);
+    /// Maximum valid value: `curve_order - 1`
+    pub const MAX: Scalar = Scalar(MAX_RAW);
+
+    /// Generates a random scalar
+    #[cfg(any(test, feature = "rand-std"))]
+    #[cfg_attr(docsrs, doc(cfg(feature = "rand-std")))]
+    pub fn random() -> Self {
+        Self::random_custom(rand::thread_rng())
+    }
+
+    /// Generates a random scalar using supplied RNG
+    #[cfg(any(test, feature = "rand"))]
+    #[cfg_attr(docsrs, doc(cfg(feature = "rand")))]
+    pub fn random_custom<R: rand::Rng>(mut rng: R) -> Self {
+        let mut bytes = [0u8; 32];
+        loop {
+            rng.fill_bytes(&mut bytes);
+            // unlikely to go past MAX
+            if let Ok(scalar) = Scalar::from_be_bytes(bytes) {
+                break scalar;
+            }
+        }
+    }
+
+    /// Tries to deserialize from big endian bytes
+    ///
+    /// **Security warning:** this function is not constant time!
+    /// Passing secret data is not recommended.
+    ///
+    /// # Errors
+    ///
+    /// Returns error when the value is above the curve order.
+    pub fn from_be_bytes(value: [u8; 32]) -> Result<Self, OutOfRangeError> {
+        // Lexicographic ordering of arrays of the same length is same as ordering of BE numbers
+        if value <= MAX_RAW {
+            Ok(Scalar(value))
+        } else {
+            Err(OutOfRangeError {})
+        }
+    }
+
+    /// Tries to deserialize from little endian bytes
+    ///
+    /// **Security warning:** this function is not constant time!
+    /// Passing secret data is not recommended.
+    ///
+    /// # Errors
+    ///
+    /// Returns error when the value is above the curve order.
+    pub fn from_le_bytes(mut value: [u8; 32]) -> Result<Self, OutOfRangeError> {
+        value.reverse();
+        Self::from_be_bytes(value)
+    }
+
+    /// Serializes to big endian bytes
+    pub fn to_be_bytes(self) -> [u8; 32] {
+        self.0
+    }
+
+    /// Serializes to little endian bytes
+    pub fn to_le_bytes(self) -> [u8; 32] {
+        let mut res = self.0;
+        res.reverse();
+        res
+    }
+
+    // returns a reference to internal bytes
+    // non-public to not leak the internal representation
+    pub(crate) fn as_be_bytes(&self) -> &[u8; 32] {
+        &self.0
+    }
+
+    pub(crate) fn as_c_ptr(&self) -> *const u8 {
+        use secp256k1_sys::CPtr;
+
+        self.as_be_bytes().as_c_ptr()
+    }
+}
+
+impl From<crate::SecretKey> for Scalar {
+    fn from(value: crate::SecretKey) -> Self {
+        Scalar(value.secret_bytes())
+    }
+}
+
+
+/// Error returned when the value of scalar is invalid - larger than the curve order.
+// Intentionally doesn't implement `Copy` to improve forward compatibility.
```

Comment:
Why does implementing `Copy` hurt forward compatibility @Kixunil?

Reply:
I had the idea of a feature gate adding boxed input. (boxed to decrease cost of moving - this is common for error types)

Reply:
Cool, cheers.